### DISCUSSION TITLE
real-time activation of ACL Mnesia on clientid and session issues

### DISCUSSION THREAD

**Discussion by goriummaximum** [emqx/emqx#6761] [Q&A]
Hi all,

I have a problem with updating clientid ACL using Mnesia while client is still in session.

After spending sometimes testing out, it seems that the newly updated ACL is activated on certain clientid only after the session with that clientid ended, or initial connect.

Is there any way to work around? Because I want to activate ACL immediately even though session with that clientid is still on (to remain data cache in message queue).

There is a use case I want to allow a certain clientid to subscribe additional topics while still maintaining the current session.

<b>Environment</b>
EMQX open source.
MQTTjs as client library.

I am looking forward for answer.
Thank you very much.

**Comment by zmstone** [comment]
The acl result is cached.
Try force cache eviction 
https://docs.emqx.io/en/broker/v4.3/advanced/cli.html#acl-command

**Comment by goriummaximum** [comment]
hi @zmstone,

Your answer is understandable, thank you.

However, I have tried the case that clearing ACL cache of a specific clientid before every time I update the ACL by <b>deleting</b> subscription access of that clientid in Mnesia and I it did not work, too. That clientid can still subscribe to the deny topic until I unsubscribe or clean up the session.

Luckily, I have found out that although I unsub that topic, the newly packets still be cached there for the session. When the client sub to that topic again, I can still get all the un-ACK data in message queue.

In the case of <b>adding</b> sub ACL on a new topic, the client follow the ACL rules in real-time.

**Comment by zmstone** [comment]
I see.
Once a client has successfully subscribed to a topic, the changes in acl does not affect it anymore.

@HJianBo when cache clean is done per client, it should perhaps trigger a recheck for subscription list?
Also, a re-subscribe (sub again without unsub)  should trigger acl check?

> Luckily, I have found out that although I unsub that topic, the newly packets still be cached there for the session. When the client sub to that topic again, I can still get all the un-ACK data in message queue.

This is perhaps due to race, we should probably not rely on this behavior.

**Reply by zmstone** [reply]
@goriummaximum have you tried re-sub (Sub without unsub)?

**Reply by goriummaximum** [reply]
@zmstone I have try re-sub everytime reconnecting but it didn't work either. ACL behavior only happen if I unsub (have just try with unsub on EMQX dashboard GUI), then sub again.

Here is my test flow for you to see more details.
1. set ACL to deny all client on topic '#'.
2. set ACL to allow client 'puber' to pub topic 'devices' (publish sample data).
3. set ACL to allow client 'suber' to sub topic 'devices'.
4. let 'suber' connect and sub to 'devices' (get data at this time).
5. disconnect 'suber'.
6. delete ACL allowing 'suber' sub to topic 'devices'.
7. clear ACL cache.
8. <b>'suber' reconnect and re-sub (stil get data).</b>
9. disconnect 'suber'.
10. unsub 'devices' of 'suber'.
11. <b>'suber' reconnect and sub. (not get data anymore, assume ACL is activated).</b>

p/s: here is one more thing that I just notice after finish step 8 is that after delete the ACL, 'suber' cannot get retain message anymore (but still get all new data from 'devices').

**Reply by zmstone** [reply]
thank you for the detailed steps.
Which EMQ X version ?

**Reply by goriummaximum** [reply]
EMQ X broker open source 4.3.11. I got from Docker.
